nupic/research/frameworks/vernon/experiments/self_supervised_experiment.py

```diff
@@ -0,0 +1,452 @@
+            def forward(self, x):
+                x = ...
+                return x
+
+            # the encode function is used during supervised training
+            def encode(self, x):
+                encoded = ...
+                return encoded
+
+    The reason for this requirement is that the loss for the unsupervised training
+    can (and usually does) operate on a different output than the representation
+    that is used during supervised training. For example, an autoencoder's forward()
+    should reconstruct its input, while its encode() function should yield the
+    intermediate representation of a typically lower dimension.
+
+    The model passed into this experiment under the name 'classifier' should take
+    the encoding provided by .encode() as input and output a classification in its
+    forward() method.
+
+    The validation step trains a new classifier on top of the existing frozen model,
+    and then proceeds to test this classifier on the test set. The number of
+    supervised training epochs to train for each validation is given by
+    supervised_training_epochs_per_validation.
+    """
+
+    def setup_experiment(self, config):
+        """
+        Configure the experiment for training.
+        :param config: Dictionary containing the configuration parameters
+            - supervised_batch_size: Supervised training batch size
+            - classifier_model_class: Classifier model class. Must inherit from
+                                      "torch.nn.Module"
+            - classifier_model_args: model model class arguments passed to the
+                                     constructor
+            - classifier_init_batch_norm: Whether or not to Initialize running batch
+                                          norm mean to 0.
+            - classifier_checkpoint_file: if not None, will start from this model. The
+                                          model must have the same model_args and
+                                          model_class as the current experiment.
+            - classifier_load_checkpoint_args_classifier: args to be passed to
+                                               `load_state_from_checkpoint`
+            - classifier_optimizer_class: Optimizer class for the optimizer.
+                                          Must inherit from "torch.optim.Optimizer"
+            - classifier_optimizer_args: Optimizer class arguments passed to the
+                                         constructor for the classifier optimizer
+            - classifier_lr_scheduler_class: Learning rate scheduler class.
+                                 Must inherit from "_LRScheduler"
+            - classifier_lr_scheduler_args: Learning rate scheduler class class
+                                            arguments passed to the constructor
+            - classifier_lr_scheduler_step_every_batch: Whether to step the lr-scheduler
```

Comment:
Minor, the indentation is off here as well as for other parameters. In this case, the `Whether ` and `after` should be aligned.